src/horizontal-pod-autoscaler.ts

```diff
@@ -0,0 +1,866 @@
+   * * Deployment
+   * * ReplicaSet
+   * * ReplicationController
+   * * StatefulSet
+   *
+   * Cannot be used with workloads that do not scale, such as DaemonSet.
+   */
+  public readonly target: Workload;
+  /**
+   * The maximum number of replicas that can be scaled up to.
+   */
+  public readonly maxReplicas: number;
+  /**
+   * The minimum number of replicas that can be scaled down to.
+   *
+   * Can be set to 0 if the alpha feature gate `HPAScaleToZero` is enabled and
+   * at least one Object or External metric is configured.
+   *
+   * @default 1
+   */
+  public readonly minReplicas?: number;
+  /**
+   * The metric conditions that trigger a scale up or scale down.
+   *
+   * @default
+   * No metrics are configured. Only scales on the requests or limits defined on the target.
+   */
+  public readonly metrics?: Metric[];
+  /**
+   * The scaling behavior when scaling up.
+   *
+   * @default
+   * Is the higher of:
+   * * Increase no more than 4 pods per 60 seconds
+   * * Double the number of pods per 60 seconds
+   */
+  public scaleUp: ScalingRules;
+  /**
+   * The scaling behavior when scaling down.
+   *
+   * @default
+   * Scale down to minReplica count with a 5 minute stabilization window.
+   */
+  public scaleDown: ScalingRules;
+
+  // Defaults
+  private readonly _defaultScaleUp = {
+    strategy: ScalingStrategy.MAX_CHANGE,
+    stabilizationWindow: Duration.seconds(0),
+    policies: [
+      {
+        replicas: Replicas.absolute(4),
+        duration: Duration.minutes(1),
+      },
+      {
+        replicas: Replicas.percent(200),
+        duration: Duration.minutes(1),
+      },
+    ],
+  };
+  private readonly _defaultScaleDown = {
+    strategy: ScalingStrategy.MAX_CHANGE,
+    stabilizationWindow: Duration.minutes(5),
+    policies: [
+      {
+        replicas: Replicas.absolute(this.minReplicas ?? 1),
+        duration: Duration.minutes(5),
+      },
+    ],
+  };
+  private readonly _defaultScalingPolicy = {
+    duration: Duration.seconds(15),
+  };
+
+
+  constructor(scope: Construct, id: string, props: HorizontalPodAutoscalerProps) {
+    super(scope, id);
+    this.apiObject = new k8s.KubeHorizontalPodAutoscalerV2(this, 'Resource', {
+      metadata: props.metadata,
+      spec: Lazy.any({ produce: () => this._toKube() }),
+    });
+
+    if (props.target.kind === 'DaemonSet') {
+      throw new Error('HorizontalPodAutoscaler cannot be used with workloads that do not scale, such as a DaemonSet.');
+    }
+    const hasResourceConstraints = props.target.containers.some((c) => this._hasRequestsOrLimits(c));
+    if (!hasResourceConstraints && !props.metrics) {
+      throw new Error('Every container in the HorizontalPodAutoscaler target must have CPU or memory resources defined');
+    }
+    if (props?.minReplicas && props.minReplicas > props.maxReplicas) {
+      throw new Error("'minReplicas' must be less than or equal to maxReplicas in order for HorizontalPodAutoscaler to scale.");
```

Comment:
Mention the violating values in the error message. I usually use parentheses:

 ```suggestion
      throw new Error(`'minReplicas' (${props.minReplicas}) must be less than or equal to 'maxReplicas' *(${props.maxReplicas}) in order for HorizontalPodAutoscaler to scale.);
```